### ISSUE TITLE
FLAGS response with only one flag(i.e. "FLAGS (\Seen)") is not taken in IMAP4Client.__cbSelect

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4069]
|<img alt="natta's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| natta reported|
|-|-|
|Trac ID|trac#4069|
|Type|defect|
|Created|2009-11-05 15:33:20Z|
|Branch|https://github.com/twisted/twisted/tree/imap4-select-tests-4069|

The prolem is about this part of IMAP4Client.__cbSelect 
```
def __cbSelect(self, (lines, tagline), rw):
        # In the absense of specification, we are free to assume:
        #   READ-WRITE access
        datum = {'READ-WRITE': rw}
        lines.append(tagline)
        for parts in lines:
            split = parts.split()
            if len(split) == 2:
                if split[1].upper().strip() == 'EXISTS':
                    try:
                        datum['EXISTS'] = int(split[0])
                    except ValueError:
                        raise IllegalServerResponse(parts)
                elif split[1].upper().strip() == 'RECENT':
                    try:
                        datum['RECENT'] = int(split[0])
                    except ValueError:
                        raise IllegalServerResponse(parts)
                else:
                    log.err('Unhandled SELECT response (1): ' + parts)
            elif split[0].upper().strip() == 'FLAGS':
```
etc.
When parts variable is equal to  "FLAGS (\Seen)" for example ( for * FLAGS (\Seen) response to SELECT command),the length of split list will be 2, but we check only for RECENT or EXISTS responses in case of length equal to 2, so our FLAGS response will be not taken into account.

The solution for this problem is very simple:
the branch
```
 elif split[0].upper().strip() == 'FLAGS':
     split = parts.split(None, 1)
     datum['FLAGS'] = tuple(parseNestedParens(split[1])[0])
```
should be also added to the if operator in the length = 2 part 

<details><summary>Searchable metadata</summary>

```
trac-id__4069 4069
type__defect defect
reporter__natta natta
priority__normal normal
milestone__ 
branch__branches_imap4_select_tests_4069 branches/imap4-select-tests-4069
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1257435200000000 1257435200000000
changetime__1270125986000000 1270125986000000
version__None None
owner__ 
cc__exarkun cc__khorn
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to natta|
|-|-|

Hi, thanks for reporting this.  I wonder if you could help out by adding a unit test which fails because of this problem?  That would make it much easier to understand the case that's not handled properly.

Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="natta's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|natta set status to `closed`|
|-|-|

In new version (9.0.0) this bug is already fixed

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

This can't really be fixed, since there's still no unit test for the behavior.  Like as not, the next release of Twisted will re-introduce whatever problem this ticket is supposed to be describing.  That kind of accidental change is what happens when there's no unit tests for correct behavior.

So it'd be really great if you could provide a unit test for this anyway.

**Comment by twisted-trac** [comment]
|<img alt="natta's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|natta commented|
|-|-|

OK, I will try to do it asap

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_6"></a>|@khorn set status to `reopened`|
|-|-|

If it's not fixed it should be an open ticket, yes?

We can close it again after an appropriate test is applied.

**Comment by twisted-trac** [comment]
|<img alt="natta's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|natta commented|
|-|-|

Replying to [khorn](#note_6):
> If it's not fixed it should be an open ticket, yes?
> 
> We can close it again after an appropriate test is applied.

Ok, so I suggest to add the following test in the test_imap.py in IMAP4ServerTestCase:
```
def testSelectOneFlag(self):
        SimpleServer.theAccount.addMailbox('test-mailbox')
        mbox = SimpleServer.theAccount.mailboxes['TEST-MAILBOX']
        mbox.flags  = ('\Flag1')
        self.selectedArgs = None
        def login():
            return self.client.login('testuser', 'password-test')
        def select():
            def selected(args):
                self.selectedArgs = args
                self._cbStopClient(None)
            d = self.client.select('test-mailbox')
            d.addCallback(selected)
            return d

        d1 = self.connected.addCallback(strip(login))
        d1.addCallback(strip(select))
        d1.addErrback(self._ebGeneral)
        d2 = self.loopback()
        return defer.gatherResults([d1, d2]).addCallback(self._cbTestSelectOneFlag)

    def _cbTestSelectOneFlag(self, ignored):
        self.assertTrue('FLAGS' in self.selectedArgs.keys())
```

**Comment by twisted-trac** [comment]
|<img alt="natta's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|natta commented|
|-|-|

Replying to [natta](#note_7):
> Replying to [khorn](#note_6):
> > If it's not fixed it should be an open ticket, yes?
> > 
> > We can close it again after an appropriate test is applied.
> 
> Ok, so I suggest to add the following test in the test_imap.py in IMAP4ServerTestCase:
> ```
> def testSelectOneFlag(self):
>         SimpleServer.theAccount.addMailbox('test-mailbox')
>         mbox = SimpleServer.theAccount.mailboxes['TEST-MAILBOX']
>         mbox.flags  = ('\Flag1')
>         self.selectedArgs = None
>         def login():
>             return self.client.login('testuser', 'password-test')
>         def select():
>             def selected(args):
>                 self.selectedArgs = args
>                 self._cbStopClient(None)
>             d = self.client.select('test-mailbox')
>             d.addCallback(selected)
>             return d
> 
>         d1 = self.connected.addCallback(strip(login))
>         d1.addCallback(strip(select))
>         d1.addErrback(self._ebGeneral)
>         d2 = self.loopback()
>         return defer.gatherResults([d1, d2]).addCallback(self._cbTestSelectOneFlag)
> 
>     def _cbTestSelectOneFlag(self, ignored):
>         self.assertTrue('FLAGS' in self.selectedArgs.keys())
> ```

Sorry, not ('\Flag1'), but ('\Flag1',). Failed for that Python feature again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

The branch for [#10722](https://github.com/twisted/twisted/issues/10722) is what fixed the behavior, it seems.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Thanks for supplying the test.  Once I tracked down the revision that fixed the problem it made it clear how the bug actually got fixed.  Since there are a bunch of `examine` tests, I think it makes sense to apply all of these to `select`.  I'm going to do that and put it in a branch for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [28762]) Branching to 'imap4-select-tests-4069'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [28763]) Put the existing IMAP4Client.examine tests into a mixin and apply them to both IMAP4Client.examine and IMAP4Client.select

refs [#4069](https://github.com/twisted/twisted/issues/4069)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set owner to @exarkun|
|-|-|

Just one comment:
 * the `_examine` method should be renamed, and the docstring fixed.

Please merge once fixed!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [28802]) Merge imap4-select-tests-4069

Author: exarkun
Reviewer: therve
Fixes: [#4069](https://github.com/twisted/twisted/issues/4069)

Apply an existing test case full of `IMAP4Client.examine` tests
to `IMAP4Client.select` as well.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
